### PR TITLE
[onert] DRAFT: Make Chrome Tracing handle multiple subgraphs

### PR DESCRIPTION
**Pull request by hyunsik-yoon**
DRAFT/WIP

Parent issue: #4901

Currently Chrome Tracing works only with one subgraphs.
This will make Chrome Tracing handle multiple subgraphs.

Signed-off-by: Hyun Sik Yoon <hyunsik.yoon.1024@gmail.com>

### CODE REVIEW

File: runtime/onert/core/include/compiler/LoweredGraph.h [hunk 1]
```diff
@@ -36,7 +36,8 @@ namespace compiler
 class LoweredGraph
 {
 public:
-  LoweredGraph(const ir::Graph &graph, const compiler::CompilerOptions &options);
+  LoweredGraph(const ir::Graph &graph, const compiler::CompilerOptions &options,
+               std::shared_ptr<util::TracingCtx> tracing_ctx);
```
Comment (wateret) [inline runtime/onert/core/include/compiler/LoweredGraph.h:40]: What about putting `tracing_ctx` in `CompilerOptions`?

File: runtime/onert/core/src/compiler/ExecutorFactory.cc [hunk 2]
```diff
@@ -241,7 +242,8 @@ void ExecutorFactory::prepareMigrantTensors(compiler::LoweredGraph &lowered_grap
 exec::IExecutor *
 ExecutorFactory::createLinearExecutor(std::unique_ptr<compiler::LoweredGraph> lowered_graph,
                                       const compiler::CompilerOptions &options,
-                                      const std::shared_ptr<exec::ExecutorMap> &executor_map)
+                                      const std::shared_ptr<exec::ExecutorMap> &executor_map,
+                                      std::shared_ptr<util::TracingCtx> tracing_ctx)
```
Comment (wateret) [inline runtime/onert/core/src/compiler/ExecutorFactory.cc:246]: https://github.com/Samsung/ONE/pull/4903/files#r531345125

File: runtime/onert/api/src/nnfw_api_internal.h [hunk 3]
```diff
@@ -156,6 +157,8 @@ struct nnfw_session
   std::unique_ptr<onert::compiler::Compiler> _compiler;
   std::shared_ptr<onert::exec::Execution> _execution;
   std::shared_ptr<onert::frontend::custom::KernelRegistry> _kernel_registry;
+
+  std::shared_ptr<onert::util::TracingCtx> _tracing_ctx;
```
Comment (wateret) [inline runtime/onert/api/src/nnfw_api_internal.h:161]: Looks like `unique_ptr` is enough (raw pointers for the other references)

File: runtime/onert/core/include/util/TracingCtx.h [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    return find == _subgraph_indices.end() ? getDefaultSubgraphIndex() : find->second;
+  }
+
+  /**
+   * @brief Get the Default Subgraph Index object. Call this when there is no subgraph index
+   *        information or when TracingCtx object is nullptr
+   *
+   * @return const ir::SubgraphIndex
+   */
+  static const ir::SubgraphIndex getDefaultSubgraphIndex() { return ir::SubgraphIndex{0}; }
```
Comment (wateret) [inline runtime/onert/core/include/util/TracingCtx.h:72]:
```suggestion
  static const ir::SubgraphIndex getPrimarySubgraphIndex() { return ir::SubgraphIndex{0}; }
```

For consistency :)

File: runtime/onert/core/include/util/TracingCtx.h [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+  // create and store unique session id managed by this class
+  void makeSessionId()
+  {
+    std::unique_lock<std::mutex> lock{_session_id_mutex};
+
+    static uint32_t session_id = 0;
+    _session_id = session_id++;
+
+    lock.unlock();
+  }
```
Comment (wateret) [inline runtime/onert/core/include/util/TracingCtx.h:46]: What about...

- Have a static `next_session_id` variable(make the mutex be for this variable)
    - `_session_id` stays
- Make a constructor that does something like `makeSessionId`(but returns ID, and assign it to `_session_id`
    - Remove public method `makeSessionId()`

File: runtime/onert/core/include/util/TracingCtx.h [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+public:
+  // create and store unique session id managed by this class
+  void makeSessionId()
+  {
+    std::unique_lock<std::mutex> lock{_session_id_mutex};
+
+    static uint32_t session_id = 0;
+    _session_id = session_id++;
+
+    lock.unlock();
```
Comment (wateret) [inline runtime/onert/core/include/util/TracingCtx.h:45]:
```suggestion
```

`unique_lock` works with RAII,  so this is redundant.

File: runtime/onert/core/include/compiler/LoweredGraph.h (hunk 1 above)
Comment (hyunsik-yoon) [reply runtime/onert/core/include/compiler/LoweredGraph.h:40]: Do you mean the following?

1. creating `TracingCtx` at `nnfw_create_session()` (as in this PR)
2. pass `TracingCtx` to `compilerOption` and let `compilerOption` has ownership of `TracingCtx`

File: runtime/onert/core/include/util/TracingCtx.h (hunk 6 above)
Comment (hyunsik-yoon) [reply runtime/onert/core/include/util/TracingCtx.h:45]: I love comment like this. :-)
(Comment that makes me study more!)

File: runtime/onert/core/include/compiler/LoweredGraph.h (hunk 1 above)
Comment (hyunsik-yoon) [reply runtime/onert/core/include/compiler/LoweredGraph.h:40]: While rewriting this draft, I decided to put `TracingCtx` into `nnfw_session` (not in `CompilerOptions`)
But with your different comment(https://github.com/Samsung/ONE/pull/4903#discussion_r531352334)
I think what you meant is
1) Let `nnfw_session` has ownership of `TracingCtx`
2) Store ref of `TracingCtx` also into `CompilerOptions` to reduce the number of params.

I will fix like the above.

Comment (hyunsik-yoon) [reply runtime/onert/core/include/compiler/LoweredGraph.h]: fixed as described in the above comment.

File: runtime/onert/core/src/compiler/ExecutorFactory.cc (hunk 2 above)
Comment (hyunsik-yoon) [reply runtime/onert/core/src/compiler/ExecutorFactory.cc]: fixed. thanks.

File: runtime/onert/api/src/nnfw_api_internal.h (hunk 3 above)
Comment (hyunsik-yoon) [reply runtime/onert/api/src/nnfw_api_internal.h]: fixed. thanks.

File: runtime/onert/core/include/util/TracingCtx.h (hunk 5 above)
Comment (hyunsik-yoon) [reply runtime/onert/core/include/util/TracingCtx.h:44]: fixed. thanks.

File: runtime/onert/core/include/util/TracingCtx.h (hunk 6 above)
Comment (hyunsik-yoon) [reply runtime/onert/core/include/util/TracingCtx.h]: fixed.

**Comment by hyunsik-yoon** [comment]
Multiple sessions are supported now:
image

**Comment by hyunsik-yoon** [comment]
All merged.

#### _How to test_
- Please go see https://github.com/Samsung/ONE/issues/4901#issuecomment-752309077

**Pull request closed without merging**